Begin fixing all the broken doctests in `compiler/`

Elliot-Roberts opened the pull request:
Begins to fix #95994.
All of them pass now but 24 of them I've marked with `ignore HELP (<explanation>)` (asking for help) as I'm unsure how to get them to work / if we should leave them as they are.
There are also a few that I marked `ignore` that could maybe be made to work but seem less important.
Each `ignore` has a rough "reason" for ignoring after it parentheses, with 

- `(pseudo-rust)` meaning "mostly rust-like but contains foreign syntax" 
- `(illustrative)` a somewhat catchall for either a fragment of rust that doesn't stand on its own (like a lone type), or abbreviated rust with ellipses and undeclared types that would get too cluttered if made compile-worthy.
- `(not-rust)` stuff that isn't rust but benefits from the syntax highlighting, like MIR.
- `(internal)` uses `rustc_*` code which would be difficult to make work with the testing setup.

Those reason notes are a bit inconsistently applied and messy though. If that's important I can go through them again and try a more principled approach. When I run `rg '```ignore \(' .` on the repo, there look to be lots of different conventions other people have used for this sort of thing. I could try unifying them all if that would be helpful.

I'm not sure if there was a better existing way to do this but I wrote my own script to help me run all the doctests and wade through the output. If that would be useful to anyone else, I put it here: https://github.com/Elliot-Roberts/rust_doctest_fixing_tool

rust-highfive commented:
Some changes occured to the CTFE / Miri engine

cc @rust-lang/miri

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @compiler-errors (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

Elliot-Roberts commented:
r? @jyn514

Elliot-Roberts commented:
Here are some of the `ignore`s that I thought maybe could be fixed (the line numbers might be slightly off):
```
compiler/rustc_middle/src/ty/util.rs:970
compiler/rustc_middle/src/ty/walk.rs:37  
compiler/rustc_span/src/hygiene.rs:801  
compiler/rustc_typeck/src/collect/type_of.rs:533        
compiler/rustc_typeck/src/outlives/implicit_infer.rs:239
```

jyn514 commented on compiler/rustc_incremental/src/assert_dep_graph.rs:
````diff
@@ -22,7 +22,8 @@
 //!
-//! ```
+//! ```ignore HELP (can this be made to work? (well, get the desired error instead of unrelated ones I mean))
+//! # #![feature(rustc_attrs)]
````
hmm, I'm not aware of any way to pass custom flags to the rustc invoked by rustdoc unfortunately :(

jyn514 commented on compiler/rustc_infer/src/infer/opaque_types.rs:
````diff
@@ -309,14 +309,21 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
     /// For example, if we had a function like this:
     ///
-    /// ```rust
+    /// ```ignore HELP (why doesn't this work?)
````
https://github.com/rust-lang/rust/issues/85433
The workaround is to add `fn main() {}` to the top.
```suggestion
    /// ```
    /// # fn main() {}
```

jyn514 commented on compiler/rustc_infer/src/infer/opaque_types.rs:
````diff
@@ -602,17 +609,17 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
 ///
 /// Example:
-/// ```rust
+/// ```ignore HELP (why does this not work?)
````
this might just be a bug??
```
error[E0308]: mismatched types
  --> src/infer/opaque_types.rs:623:16
   |
7  |         pub type Baz = impl Bar;
   |                        -------- the expected opaque type
...
12 |     fn f2() -> bar::Baz { /* ... */ }
   |        --      ^^^^^^^^ expected opaque type, found `()`
   |        |
   |        implicitly returns `()` as its body has no tail or `return` expression
   |
   = note: expected opaque type `Baz`
                found unit type `()`
```

jyn514 commented on compiler/rustc_middle/src/mir/mod.rs:
````diff
@@ -746,27 +746,30 @@ pub enum BorrowKind {
-    ///     let y = (&mut Env { &x }, fn_ptr);  // Closure is pair of env and fn
-    ///     fn fn_ptr(env: &mut Env) { **env.x += 5; }
-    ///
+    /// ```ignore HELP (this does indeed error but I'm not sure it's the right one?)
+    /// struct Env { x: & &mut isize }
````
```suggestion
    /// ```compile_fail,E0594
    /// struct Env<'a> { x: &'a &'a mut isize }
```
and also change the initialization to `x: &x` on line 760.

jyn514 commented on compiler/rustc_middle/src/mir/mod.rs:
````diff
@@ -746,27 +746,30 @@ pub enum BorrowKind {
-    ///     let y = (&mut Env { &mut x }, fn_ptr); // changed from &x to &mut x
-    ///     fn fn_ptr(env: &mut Env) { **env.x += 5; }
-    ///
+    /// ```ignore HELP (it seems like it's a syntax issue but I don't understand what's going on well enough to say)
````
same changes as above

jyn514 commented on compiler/rustc_middle/src/mir/mod.rs:
````diff
@@ -1004,7 +1007,7 @@ pub struct LocalDecl<'tcx> {
     /// as if the let-statements were `match` expressions:
     ///
-    /// ```
+    /// ```ignore HELP (compiler suggests annotating the parse() with the type (u32). is that a good idea?)
````
yes, that seems fine

jyn514 commented on compiler/rustc_middle/src/traits/mod.rs:
````diff
@@ -47,7 +47,8 @@ pub enum Reveal {
     ///
-    /// ```
+    /// ```ignore HELP (is E0575 the right error for this?)
+    /// #![feature(specialization)]
````
no, the comment originally forgot a variable:
```diff
diff --git a/compiler/rustc_middle/src/traits/mod.rs b/compiler/rustc_middle/src/traits/mod.rs
index da418eb1355..ffde1294ec6 100644
--- a/compiler/rustc_middle/src/traits/mod.rs
+++ b/compiler/rustc_middle/src/traits/mod.rs
@@ -47,7 +47,7 @@ pub enum Reveal {
     /// impl. Concretely, that means that the following example will
     /// fail to compile:
     ///
-    /// ```ignore HELP (is E0575 the right error for this?)
+    /// ```compile_fail,E0308
     /// #![feature(specialization)]
     /// trait Assoc {
     ///     type Output;
@@ -58,7 +58,7 @@ pub enum Reveal {
     /// }
     ///
     /// fn main() {
-    ///     let <() as Assoc>::Output = true;
+    ///     let x: <() as Assoc>::Output = true;
     /// }
     /// ```
     UserFacing,
```

jyn514 commented on compiler/rustc_middle/src/ty/closure.rs:
````diff
@@ -373,16 +373,18 @@ pub enum BorrowKind {
     /// form, we'd encounter an error with the code as written:
     ///
-    /// ```
+    /// ```ignore HELP (same as others, expected identifier, found & while parsing Env { &x })
````
same fix as before

jyn514 commented on compiler/rustc_middle/src/ty/closure.rs:
````diff
@@ -391,9 +393,10 @@ pub enum BorrowKind {
     /// an `&mut` borrow:
     ///
-    /// ```
+    /// ```ignore HELP (same as others)
````
same fix as before

jyn514 commented on compiler/rustc_middle/src/ty/inhabitedness/mod.rs:
````diff
@@ -56,7 +56,8 @@ impl<'tcx> TyCtxt<'tcx> {
     /// # Example
-    /// ```rust
+    /// ```ignore HELP (should I just add 'use super::Void;' to 'mod c' or does that defeat the point of the example?)
+    /// #![feature(never_type)]
````
"visibly" here is referring to exhaustiveness, not name resolution. `use super::Void` should be fine.

jyn514 commented on compiler/rustc_middle/src/ty/sty.rs:
````diff
@@ -187,12 +187,14 @@ pub enum TyKind<'tcx> {
     /// may end up as something like `for<'a> [Vec<i32>, &'a Vec<i32>]`:
     ///
-    /// ```rust
+    /// ```ignore HELP (is this supposed to error? the code looks purpose built to trigger E0626 but I can't tell if that's what the doc wants?)
````
no, I don't think it's supposed to error - moving the `yield a;` a line earlier seems fine. but @compiler-errors would probably know better than me.

jyn514 commented on compiler/rustc_mir_build/src/build/expr/as_operand.rs:
````diff
@@ -42,15 +42,16 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
     ///
-    /// ```rust
-    /// fn foo(p: dyn Debug) { ... }
+    /// ```ignore HELP (should that p in bar be box_p?)
````
yes. but it still fails to compile after - maybe the original code meant to use `foo(&*box_p)` instead? `dyn Debug` can't be passed to a function.

(either that, or this is left over from some pre-1.0 code.)

Elliot-Roberts replied:
I added `#![feature(unsized_locals, unsized_fn_params)]` and it worked. Does that seem like a bad fix?

jyn514 commented on compiler/rustc_monomorphize/src/collector.rs:
````diff
@@ -91,7 +91,8 @@
 //! an edge. Consider the following example:
 //!
-//! ```rust
+//! ```ignore HELP (the error suggests changing to f: &dyn Fn(i32), does that seem ok?)
````
yes. This is a 2021 edition change, it's just syntactic.

jyn514 commented on compiler/rustc_resolve/src/late/lifetimes.rs:
````diff
@@ -406,7 +406,7 @@ pub fn provide(providers: &mut ty::query::Providers) {
 /// be a cycle. Consider this example:
 ///
-/// ```rust
+/// ```ignore HELP (compiler suggests adding another lifetime variable to Sub. why?)
````
`Sub` is declared with a lifetime parameter, so it needs to be defined with one as well. Not sure what it should be. @jackh726 may know.

jackh726 replied:
I think would be
```rust
trait Sub<'b>: for<'a> Base<'a> {
    type SubItem: for<'a> Sub<'a, BaseItem = &'b u32>;
}
```

jackh726 replied:
Or just
```rust
trait Sub<'b>: for<'a> Base<'a> {
    type SubItem: Sub<'b, BaseItem = &'b u32>;
}
```

Not sure, not entirely sure what this is supposed to be doing

jyn514 commented on compiler/rustc_span/src/hygiene.rs:
````diff
@@ -756,7 +757,8 @@ impl SyntaxContext {
     /// For example:
     ///
-    /// ```rust
+    /// ```ignore HELP (this fails saying f cannot be found. is that intentional?)
````
yes, see the comment "this is a resolution error" below.

jyn514 commented on compiler/rustc_typeck/src/check/mod.rs:
````diff
@@ -884,16 +884,19 @@ fn report_unexpected_variant_res(tcx: TyCtxt<'_>, res: Res, span: Span) {
-///
-///     f(1, 2);
-///
+/// ```ignore HELP (is there a way to set TupleArgumentsFlag so that this compiles?)
````
I think this is used for `Fn*` traits. Not sure how to replicate in user code.

jyn514 commented on compiler/rustc_typeck/src/check/regionck.rs:
````diff
@@ -269,7 +269,7 @@ impl<'a, 'tcx> RegionCtxt<'a, 'tcx> {
     /// Consider this silly example:
     ///
-    /// ```
+    /// ```ignore HELP (is that @ intentional?)
````
`@` is *old* syntax for "garbage-collected pointer": https://pcwalton.github.io/2013/06/02/removing-garbage-collection-from-the-rust-language.html. Not sure how to keep the same semantics - maybe `Box<i32>` is close?

jyn514 commented on compiler/rustc_typeck/src/check/upvar.rs:
````diff
@@ -447,7 +447,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// Eg:
-    /// ```rust,no_run
+    /// ```ignore HELP ("could not find Display in fmt", is there any way to make this work? I assume we don't have std at this point?)
+    /// #[derive(std::fmt::Display)]
````
There is no `Display` derive macro. Only `Debug` can be derived.

Elliot-Roberts replied:
Right :facepalm:

jyn514 commented on compiler/rustc_typeck/src/check/upvar.rs:
````diff
@@ -1319,7 +1320,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// Consider another example:
     ///
-    /// ```rust
+    /// ```ignore HELP (what is that move(y.0)?)
````
I think that's MIR and can't be written in user code.

jyn514 commented on compiler/rustc_typeck/src/check/upvar.rs:
````diff
@@ -2213,7 +2214,10 @@ fn determine_place_ancestry_relation<'tcx>(
 /// Reason we only drop the last deref is because of the following edge case:
 ///
-/// ```rust
+/// ```ignore HELP (I feel like this can be made to work but I'm doing something wrong)
````
try this:
```diff
diff --git a/compiler/rustc_typeck/src/check/upvar.rs b/compiler/rustc_typeck/src/check/upvar.rs
index aef6c2b68ab..d51588d2737 100644
--- a/compiler/rustc_typeck/src/check/upvar.rs
+++ b/compiler/rustc_typeck/src/check/upvar.rs
@@ -2214,8 +2214,8 @@ fn determine_place_ancestry_relation<'tcx>(
 ///
 /// Reason we only drop the last deref is because of the following edge case:
 ///
-/// ```ignore HELP (I feel like this can be made to work but I'm doing something wrong)
-/// # struct A { field_of_a: i32 }
+/// ```
+/// # struct A { field_of_a: Box<i32> }
 /// # struct B {}
 /// # struct C<'a>(&'a i32);
 /// struct MyStruct<'a> {
@@ -2225,7 +2225,7 @@ fn determine_place_ancestry_relation<'tcx>(
 /// }
 ///
 /// fn foo<'a, 'b>(m: &'a MyStruct<'b>) -> impl FnMut() + 'static {
-///     let c = || drop(&*m.a.field_of_a);
+///     || drop(&*m.a.field_of_a)
 ///     // Here we really do want to capture `*m.a` because that outlives `'static`
 ///
 ///     // If we capture `m`, then the closure no longer outlives `'static'
```

jyn514 reviewed (commented):
> Those reason notes are a bit inconsistently applied and messy though. If that's important I can go through them again and try a more principled approach. When I run rg '```ignore \(' . on the repo, there look to be lots of different conventions other people have used for this sort of thing. I could try unifying them all if that would be helpful.

I don't think it's too important to clean these up, the categories you have seem fine :+1: no need to change existing tests.

I left suggestions for most of your `HELP` comments :)

> (internal) uses rustc_* code which would be difficult to make work with the testing setup.

In retrospect, I think I was wrong on the issue and this should work out of the box as long as it only uses crates that have previously been compiled (see also https://github.com/rust-lang/rust/pull/95445). Maybe try running these and see if they work?

Elliot-Roberts commented on compiler/rustc_data_structures/src/tagged_ptr.rs:
````diff
@@ -45,7 +45,8 @@ pub unsafe trait Pointer: Deref {
     ///
-    /// ```rust
+    /// ```ignore HELP (should we try to make this runnable? and why does this import not seem to work?)
+    /// # use std::ops::Deref;
````
I'll try that again. I might have somehow been looking at stale error messages (oops sorry).
I used `--stage 1` for the entire thing. Which are the stages we would want to run in CI?

jyn514 replied:
CI always uses --stage 2. At some point I want to test that other stages work, but those will be tests of bootstrap itself, not that the doctests are correct.

Elliot-Roberts commented:
Thanks for all your help so far, @jyn514! I'm learning lots. 
I had hoped to fix more myself and leave less of the problem solving to you but with the iteration times so slow it probably would have taken me a while longer. (Some of them I should have just googled/rtfm for a little more though; sorry about that).

I'll have the straightforward fixes done soon and then we'll just have the handful that need expert input :)

jyn514 commented:
Thank you for tackling this!

> with the iteration times so slow it probably would have taken me a while longer

I highly recommend cherry-picking my fix for --stage 0 so the iteration time is faster. It was taking me about ~10 seconds to check each suggestion I made, I wouldn't have gone through them all one by one if it took me 10 minutes each time.

Elliot-Roberts commented:
Oh ok wow haha. I luckily only had to recompile everything <10 times by working in batches but that sounds like a way better approach.

Oh and for the ones I marked `ignore (internal)`, I of course only marked them that because they fail and adding the obvious `use rustc_...;` suggested by the compiler didn't immediately fix the problem. I can go look at them again now. There are 29 of them with 19 identical seeming ones just in `compiler/rustc_data_structures/src/owning_ref/mod.rs`.

Elliot-Roberts commented on compiler/rustc_ast_lowering/src/lib.rs:
````diff
@@ -229,7 +229,7 @@ enum ImplTraitContext<'b, 'a> {
         ///
-        /// ```
+        /// ```ignore HELP (how would you make a stub trait for this?)
         /// type A<'b> = impl for<'a> Trait<'a, Out = impl Sized + 'a>;
````
Should we ignore this one as just an "illustration" then? The original single line of code gets quite buried when we add all that. If we figure out how to make it compile we could just hide the extra lines, but it would still be harder to read the un-rendered source (is that at all a priority for us? I've been considering it moderately important as my vscode doesn't have built-in doc rendering unless I hover on an item which is tedious and doesn't work for everything, but if we expect most future readers to have rendering then I can see why we wouldn't bother with it.)

jyn514 replied:
I think it's semi important to figure out what the comment is describing in the first place - it's unclear to me now even having read it several times and played around with the example, so it will almost certainly be unclear to whoever reads it next without that context. But I agree making it compile is probably not worth it.

oli-obk replied:
It's unclear to me, too. Igot all the way to

```rust
#![feature(type_alias_impl_trait)]
trait Trait<'a> { type Out: 'a;}
impl<'a> Trait<'a> for i32 { type Out = String;}
type A = impl for<'a> Trait<'a, Out = impl Sized + 'a>;
fn foo() -> A {
    0_i32
}

```

which is just a bug that it ain't accepted.i'll openable issuefor it And try to figure out the details of what's going on with this field at the same time.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6051942569?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/89e157421d7a88549f076051a612abcd9df28d3e/checks/6051942569/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

Elliot-Roberts commented:
Alright, I fixed the `HELP`s I could and also most of the `ignore (internal)`s.
Now left to address are:

- one `ignore LATER` I left in by accident with scratch code. I think that one might need to be an `ignore (illustrative)` because the objects we need to construct to make the example work look big. Are there easier ways than giant struct literals?
- one `ignore HELP` of a new problem (and continuation of the same problem in `ignore CONT`). Also for that one, I added a dev-dependency which might be undesirable?
- seven `ignore UNSOLVED`s for previous problems we don't have solutions for yet.

bjorn3 commented on compiler/rustc_builtin_macros/src/deriving/encodable.rs:
````diff
@@ -5,14 +5,14 @@
 //! For example, a type like:
 //!
-//! ```
+//! ```ignore HELP (it says RustcEncodable and RustcEncodable are missing generic parameters? also it says these are deprecated and I can't find any instances of them outside these files. can we remove the whole thing?)
````
They are deprecated but stable and thus can't be removed. They need to be used in combination with the rustc-serialize crate on crates.io (a precursor to serde). Note that compiler/rustc_serialize in this repo has diverged significantly from what RustcEncodable and RustcDecodable need.

Elliot-Roberts replied:
Oh I didn't realize there was code in `compiler/` that made stability guarantees to code outside the rust repo.
So I should put the crates.io `rustc-serialize` in the [dev-dependencies]`?

bjorn3 replied:
I think ignoring it is fine. There are already some tests for it in `src/test` anyway.

bjorn3 commented on compiler/rustc_parse/src/parser/item.rs:
````diff
@@ -1629,7 +1629,7 @@ impl<'a> Parser<'a> {
     /// Parses a declarative macro 2.0 definition.
     /// The `macro` keyword has already been parsed.
-    /// ```
+    /// ```ignore (not-rust)
````
Maybe this should be `plain`? That way it won't get rust syntax highlighting either.

jyn514 replied:
Yeah, just `text` here is better than ignore.

Elliot-Roberts replied:
Alright, so all the grammar ones should be `text`? I thought the syntax highlighting looked ok on them personally but I can see why you wouldn't want it.

jyn514 replied:
I don't think we should mark things that aren't rust as rust, no. Anything using EBNF should be `text` for now (rustdoc currently only supports syntax highlighting for rust: https://github.com/rust-lang/rust/issues/56213).

Elliot-Roberts replied:
Oh ok so we just shouldn't mark things that aren't rust as `ignore` just for the (correct-ish looking but technically incorrect) highlighting?
The main classes of things I did that for are:

- ebnf (which I'll go make ` ```ebnf ` now)
- MIR, which I marked as `ignore (MIR)` (forgot I separated those from the other `(not-rust)` ones)
- chalk style type stuff, e.g. 
```
forall (U: Trait + ?Sized) {
    if (Self: Unsize<U>) {
        Receiver: DispatchFromDyn<Receiver[Self => U]>
    }
}
```
and
```
for<...> <T as Trait>::U == V
```
- plus a few odd ones out (shall we make them all text? am I misunderstanding some of them?)
  - https://github.com/rust-lang/rust/pull/96094/files/cd7734e958a70641c9492aa2e33572a4de4a83be#diff-5e6fdd07e3ea2b5e20c9fdd3914d1e781ecb4ac4f5349cb2ce33732da8d8a597R173-R175
  - https://github.com/rust-lang/rust/pull/96094/files/cd7734e958a70641c9492aa2e33572a4de4a83be#diff-ee45792072b570212b473de3da83a965c3ce9b74a86109d27e818f7486d8180aR13-R16
  - https://github.com/rust-lang/rust/pull/96094/files/cd7734e958a70641c9492aa2e33572a4de4a83be#diff-a75cbce2712ef8e4b4013fc2189dcd0890974a11880e02ca16a493228add9619R1896
  - https://github.com/rust-lang/rust/pull/96094/files/cd7734e958a70641c9492aa2e33572a4de4a83be#diff-1677e6785e45881b13f3477c96f0e619c7ee07cd0fc54fb91b877cf82b3df22fR9

Should I make the MIR and chalk ones ` ```mir ` and ` ```chalk ` or ` ```text` or are they ok as-is because they're rust-y enough?

jyn514 replied:
Nah, I'd say just leave them as is. The main thing is to get the tests passing.

jyn514 commented:
> one ignore LATER I left in by accident with scratch code. I think that one might need to be an ignore (illustrative) because the objects we need to construct to make the example work look big. Are there easier ways than giant struct literals?

I think it's useful to have at least one example of what `TestDescAndFn` looks like. No need to have 3 separate struct literals. We should first confirm that's what the compiler actually expands `#[test]` function to, though.

bors commented:
:umbrella: The latest upstream changes (presumably #91557) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Elliot-Roberts commented:
I think I've done most of what I know how to do and incorporated all the feedback given so far (thanks for the help everyone!).
Now there are just 7 `ignore UNSOLVED`s left to address. (ones we haven't decided how to fix yet). Each has a small explanation/question in parentheses by it.
I might try to investigate one that's a suspected bug, but at my current level of experience that could take a while haha.
If anyone has time to advise me in working on one, that would be great too. I'm not quite sure where to start.

jyn514 commented:
I think merging as-is is fine too, no need to do everything at once. We can keep the tracking issue open until we figure out how to get rid of the `ignore`.

jyn514 commented:
@compiler-errors I think this is ready for review - I don't have bors permissions.

klensy commented:
Maybe squash? Some commits add\remove the same things.

Elliot-Roberts commented:
Right, thanks for the reminder.
On a separate note, should the `ignore UNSOLVED`s stay as they are with the comments in parentheses? I had originally written them to be very temporary things.

compiler-errors commented:
I agree with @jyn514 that this is fine to merge as-is.

@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 7907385999b4a83d37ed31d334f3ed9ca02983a1 has been approved by `compiler-errors`

<!-- @bors r=compiler-errors 7907385999b4a83d37ed31d334f3ed9ca02983a1 -->
<!-- homu: {"type":"Approved","sha":"7907385999b4a83d37ed31d334f3ed9ca02983a1","approver":"compiler-errors"} -->

compiler-errors commented:
Apparently `ignore` blocks need to parse as valid Rust code, which is not the case for some of the examples here? Can you fix these @Elliot-Roberts? Thanks!

@bors r-
@bors rollup=never

Elliot-Roberts commented:
Oh ok. I think I'll have to change a few older ignores too. Like this one from 5 years ago: https://github.com/rust-lang/rust/blob/74cea9fdb9ca77d4104df5c28124817004ead6a3/compiler/rustc_ast/src/ast.rs#L1479-L1487

Elliot-Roberts commented:
Wait never mind, [rustdoc doesn't mind that one](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/struct.QSelf.html).
In the instance identified by CI, it appears the only problem was the un-closed curly brace??

So, given that the error message says 
> `ignore` code blocks require valid Rust code for syntax highlighting; mark blocks that do not contain Rust code as text

should I still mark everything that isn't valid rust as text, or should I keep the ones that rustdoc is OK with?

Elliot-Roberts commented:
Looks like there were just a couple unmatched braces in one file that were causing rustdoc to fail to parse for syntax highlighting :sweat_smile:.

Elliot-Roberts commented:
I suppose a number of them would be better as text though. I didn't realize rustdoc shows a warning saying the code isn't tested when you use ignore. That paired with rustdoc's syntax highlighting not actually doing much makes it not very worthwhile.
Here is, for example, some highlighting in my vscode:
![image](https://user-images.githubusercontent.com/32186556/167088253-2ee76530-c373-4d9a-8c34-3372bf287004.png)
versus in rustdoc:
![image](https://user-images.githubusercontent.com/32186556/167089038-ec0ef396-3b85-41d8-b66c-3fab147686c4.png)
I thought the highlighting was preferable when I saw it in vscode, but in rustdoc it seems worse than ` ```text`. Potentially even misleading?

I'd appreciate others' opinions.

compiler-errors commented:
> should I still mark everything that isn't valid rust as text, or should I keep the ones that rustdoc is OK with?

@Elliot-Roberts: I think you should mark it as text, syntax highlighting may get messed up by stray tokens even if rustdoc doesn't get angry when parsing it, and yeah, that "this example is not tested" string is a bit misleading.

... though I have been less involved in this thread compared to e.g. @jyn514 who may have more of a opinion.

jyn514 commented:
We should use `text` unless we're actively looking for syntax highlighting, yeah.

See also https://github.com/rust-lang/rust/pull/96573 which tries to make this a little less confusing.

Elliot-Roberts commented:
ok hm. reading #96573 and learning that it's expected to be *possible* to run doctests marked with `ignore` changes things a bit. So *everying* that can't compile should be text? including fragments that only fail for undeclared stuff?

I'll start by just using text for everything that wouldn't Parse. I expect there will be some grey area for even that, though.

jyn514 commented:
> ok hm. reading https://github.com/rust-lang/rust/pull/96573 and learning that it's expected to be possible to run doctests marked with ignore changes things a bit. So everying that can't compile should be text? including fragments that only fail for undeclared stuff?

Don't spend too much more time on this. If it passes CI and it looks vaguely right in the generated docs, it's fine :) no need to rewrite half your work.

compiler-errors commented:
+1, don't waste too much time on it -- I can re-approve it as soon as you can locally confirm that the tests don't fail like they do in https://github.com/rust-lang-ci/rust/runs/6316344673?check_suite_focus=true

Elliot-Roberts commented:
Ok thanks for the reassurance :)

@compiler-errors How do I test that locally? I have at least verified that rusdoc succeeds for everything in `compiler/` (https://github.com/rust-lang/rust/pull/96094/commits/647d0b6dd3008b2d9d3a57a914bd1487da9133cf was all that was needed)

And I can still go make more of them text as now it's bothering me. I have free time today to work on it.

compiler-errors commented:
Frankly.... I don't actually know. Anywho, let's try to merge this now.

> And I can still go make more of them text as now it's bothering me

Feel free to open a new PR and `r? @compiler-errors` if you want me to review more follow-up changes.

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit 647d0b6dd3008b2d9d3a57a914bd1487da9133cf has been approved by `compiler-errors`

<!-- @bors r=compiler-errors 647d0b6dd3008b2d9d3a57a914bd1487da9133cf -->
<!-- homu: {"type":"Approved","sha":"647d0b6dd3008b2d9d3a57a914bd1487da9133cf","approver":"compiler-errors"} -->

bors commented:
:hourglass: Testing commit 647d0b6dd3008b2d9d3a57a914bd1487da9133cf with merge 574830f5730f2cfc3abdb486428d33d541c0abee...
<!-- homu: {"type":"BuildStarted","head_sha":"647d0b6dd3008b2d9d3a57a914bd1487da9133cf","merge_sha":"574830f5730f2cfc3abdb486428d33d541c0abee"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6333516669?check_suite_focus=true)
Approved by: compiler-errors
Pushing 574830f5730f2cfc3abdb486428d33d541c0abee to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6333516669?check_suite_focus=true"},"merge_sha":"574830f5730f2cfc3abdb486428d33d541c0abee"} -->

rust-timer commented:
Finished benchmarking commit (574830f5730f2cfc3abdb486428d33d541c0abee): [comparison url](https://perf.rust-lang.org/compare.html?start=36aa7c143672af30bfcca01e5924b326f93fd922&end=574830f5730f2cfc3abdb486428d33d541c0abee).

**Summary**: This benchmark run did not return any relevant results.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
